diff --git a/clang-tools-extra/clang-tidy/misc/CMakeLists.txt b/clang-tools-extra/clang-tidy/misc/CMakeLists.txt
index b4cc2bcda831..0d10d978265a 100644
--- a/clang-tools-extra/clang-tidy/misc/CMakeLists.txt
+++ b/clang-tools-extra/clang-tidy/misc/CMakeLists.txt
@@ -5,6 +5,7 @@ set(LLVM_LINK_COMPONENTS
 
 add_clang_library(clangTidyMiscModule
   DefinitionsInHeadersCheck.cpp
+  MaxTokensCheck.cpp
   MiscTidyModule.cpp
   MisplacedConstCheck.cpp
   NewDeleteOverloadsCheck.cpp
diff --git a/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.cpp b/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.cpp
new file mode 100644
index 000000000000..4def2d9db7ef
--- /dev/null
+++ b/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.cpp
@@ -0,0 +1,58 @@
+//===--- MaxTokensCheck.cpp - clang-tidy ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "MaxTokensCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/PPCallbacks.h"
+#include "clang/Lex/Preprocessor.h"
+#include "llvm/Support/FormatVariadic.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace misc {
+namespace {
+class MaxTokensPPCallbacks : public PPCallbacks {
+public:
+  MaxTokensPPCallbacks(MaxTokensCheck *Check) : Check(Check) {}
+
+  void diag(SourceLocation Loc, unsigned TokenCount, unsigned MaxTokenCount) {
+    if (TokenCount > MaxTokenCount) {
+      std::string Msg = llvm::formatv(
+          "Number of tokens ({0}) exceeds the specified maximum ({1})",
+          TokenCount, MaxTokenCount);
+      Check->diag(Loc, Msg);
+    }
+  }
+
+  void PragmaMaxTokensHere(SourceLocation Loc, unsigned CurrTokenCount,
+                           unsigned MaxTokenCount) override {
+    diag(Loc, CurrTokenCount, MaxTokenCount);
+  }
+
+  void PragmaMaxTokensTotal(SourceLocation Loc, unsigned TokenCount,
+                            unsigned MaxTokenCount) override {
+    diag(Loc, TokenCount, MaxTokenCount);
+  }
+
+private:
+  MaxTokensCheck *Check;
+};
+} // namespace
+
+void MaxTokensCheck::registerPPCallbacks(const SourceManager &SM,
+                                         Preprocessor *PP,
+                                         Preprocessor *ModuleExpanderPP) {
+  PP->addPPCallbacks(std::make_unique<MaxTokensPPCallbacks>(this));
+}
+
+} // namespace misc
+} // namespace tidy
+} // namespace clang
diff --git a/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.h b/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.h
new file mode 100644
index 000000000000..b5b917e93816
--- /dev/null
+++ b/clang-tools-extra/clang-tidy/misc/MaxTokensCheck.h
@@ -0,0 +1,33 @@
+//===--- MaxTokensCheck.h - clang-tidy --------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MISC_MAXTOKENSCHECK_H
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MISC_MAXTOKENSCHECK_H
+
+#include "../ClangTidyCheck.h"
+
+namespace clang {
+namespace tidy {
+namespace misc {
+
+/// Implements the -Wmax-tokens clang flag as a check
+///
+/// For the user-facing documentation see:
+/// http://clang.llvm.org/extra/clang-tidy/checks/misc-max-tokens.html
+class MaxTokensCheck : public ClangTidyCheck {
+public:
+  MaxTokensCheck(StringRef Name, ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
+  void registerPPCallbacks(const SourceManager &SM, Preprocessor *PP, Preprocessor *ModuleExpanderPP) override;
+};
+
+} // namespace misc
+} // namespace tidy
+} // namespace clang
+
+#endif // LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MISC_MAXTOKENSCHECK_H
diff --git a/clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp b/clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp
index 72dc89e8519e..3593a360b9ad 100644
--- a/clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp
+++ b/clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp
@@ -10,6 +10,7 @@
 #include "../ClangTidyModule.h"
 #include "../ClangTidyModuleRegistry.h"
 #include "DefinitionsInHeadersCheck.h"
+#include "MaxTokensCheck.h"
 #include "MisplacedConstCheck.h"
 #include "NewDeleteOverloadsCheck.h"
 #include "NoRecursionCheck.h"
@@ -34,6 +35,8 @@ public:
   void addCheckFactories(ClangTidyCheckFactories &CheckFactories) override {
     CheckFactories.registerCheck<DefinitionsInHeadersCheck>(
         "misc-definitions-in-headers");
+    CheckFactories.registerCheck<MaxTokensCheck>(
+        "misc-max-tokens");
     CheckFactories.registerCheck<MisplacedConstCheck>("misc-misplaced-const");
     CheckFactories.registerCheck<NewDeleteOverloadsCheck>(
         "misc-new-delete-overloads");
diff --git a/clang-tools-extra/docs/ReleaseNotes.rst b/clang-tools-extra/docs/ReleaseNotes.rst
index 733f016682bd..cf5e1d251d06 100644
--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -301,6 +301,11 @@ New checks
   Finds includes of system libc headers not provided by the compiler within
   llvm-libc implementations.
 
+- New :doc:`misc-max-tokens
+  <clang-tidy/checks/misc-max-tokens>` check.
+
+  FIXME: add release notes.
+
 - New :doc:`misc-my-first-check
   <clang-tidy/checks/misc-potential-unbounded-loop>` check.
 
diff --git a/clang-tools-extra/docs/clang-tidy/checks/list.rst b/clang-tools-extra/docs/clang-tidy/checks/list.rst
index b8d9b63bedb9..93e3b012badb 100644
--- a/clang-tools-extra/docs/clang-tidy/checks/list.rst
+++ b/clang-tools-extra/docs/clang-tidy/checks/list.rst
@@ -195,6 +195,7 @@ Clang-Tidy Checks
    `llvmlibc-implementation-in-namespace <llvmlibc-implementation-in-namespace.html>`_,
    `llvmlibc-restrict-system-libc-headers <llvmlibc-restrict-system-libc-headers.html>`_, "Yes"
    `misc-definitions-in-headers <misc-definitions-in-headers.html>`_, "Yes"
+   `misc-max-tokens <misc-max-tokens.html>`_, "Yes"
    `misc-misplaced-const <misc-misplaced-const.html>`_,
    `misc-new-delete-overloads <misc-new-delete-overloads.html>`_,
    `misc-no-recursion <misc-no-recursion.html>`_,
diff --git a/clang-tools-extra/docs/clang-tidy/checks/misc-max-tokens.rst b/clang-tools-extra/docs/clang-tidy/checks/misc-max-tokens.rst
new file mode 100644
index 000000000000..e6995cd8748b
--- /dev/null
+++ b/clang-tools-extra/docs/clang-tidy/checks/misc-max-tokens.rst
@@ -0,0 +1,6 @@
+.. title:: clang-tidy - misc-max-tokens
+
+misc-max-tokens
+===============
+
+FIXME: Describe what patterns does the check detect and why. Give examples.
diff --git a/clang-tools-extra/test/clang-tidy/checkers/misc-max-tokens.cpp b/clang-tools-extra/test/clang-tidy/checkers/misc-max-tokens.cpp
new file mode 100644
index 000000000000..f492e20515c5
--- /dev/null
+++ b/clang-tools-extra/test/clang-tidy/checkers/misc-max-tokens.cpp
@@ -0,0 +1,14 @@
+// RUN: %check_clang_tidy %s misc-max-tokens %t
+
+// FIXME: Add something that triggers the check here.
+void f();
+// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: function 'f' is insufficiently awesome [misc-max-tokens]
+
+// FIXME: Verify the applied fix.
+//   * Make the CHECK patterns specific enough and try to make verified lines
+//     unique to avoid incorrect matches.
+//   * Use {{}} for regular expressions.
+// CHECK-FIXES: {{^}}void awesome_f();{{$}}
+
+// FIXME: Add something that doesn't trigger the check here.
+void awesome_f2();
diff --git a/clang/include/clang/Lex/PPCallbacks.h b/clang/include/clang/Lex/PPCallbacks.h
index de5e8eb2ca22..8351b3492837 100644
--- a/clang/include/clang/Lex/PPCallbacks.h
+++ b/clang/include/clang/Lex/PPCallbacks.h
@@ -276,6 +276,16 @@ public:
   /// is read.
   virtual void PragmaAssumeNonNullEnd(SourceLocation Loc) {}
 
+  /// TODO: MaxTokens documentation
+  virtual void PragmaMaxTokensHere(SourceLocation Loc,
+                                   unsigned CurrTokenCount,
+                                   unsigned MaxTokenCount) {}
+
+  /// TODO: MaxTokens documentation
+  virtual void PragmaMaxTokensTotal(SourceLocation Loc,
+                                   unsigned TokenCount,
+                                   unsigned MaxTokenCount) {}
+
   /// Called by Preprocessor::HandleMacroExpandedIdentifier when a
   /// macro invocation is found.
   virtual void MacroExpands(const Token &MacroNameTok,
@@ -535,6 +545,16 @@ public:
     Second->PragmaAssumeNonNullEnd(Loc);
   }
 
+  void PragmaMaxTokensHere(SourceLocation Loc, unsigned CurTokenCount, unsigned MaxTokenCount) override {
+    First->PragmaMaxTokensHere(Loc, CurTokenCount, MaxTokenCount);
+    Second->PragmaMaxTokensHere(Loc, CurTokenCount, MaxTokenCount);
+  }
+
+  void PragmaMaxTokensTotal(SourceLocation Loc, unsigned TokenCount, unsigned MaxTokenCount) override {
+    First->PragmaMaxTokensTotal(Loc, TokenCount, MaxTokenCount);
+    Second->PragmaMaxTokensTotal(Loc, TokenCount, MaxTokenCount);
+  }
+
   void MacroExpands(const Token &MacroNameTok, const MacroDefinition &MD,
                     SourceRange Range, const MacroArgs *Args) override {
     First->MacroExpands(MacroNameTok, MD, Range, Args);
diff --git a/clang/lib/Parse/ParsePragma.cpp b/clang/lib/Parse/ParsePragma.cpp
index 6402b31d00b2..4fb081aa5618 100644
--- a/clang/lib/Parse/ParsePragma.cpp
+++ b/clang/lib/Parse/ParsePragma.cpp
@@ -3491,6 +3491,11 @@ void PragmaMaxTokensHereHandler::HandlePragma(Preprocessor &PP,
     return;
   }
 
+  PPCallbacks *Callbacks = PP.getPPCallbacks();
+  if (Callbacks) {
+    Callbacks->PragmaMaxTokensHere(Loc, PP.getTokenCount(), MaxTokens);
+  }
+
   if (PP.getTokenCount() > MaxTokens) {
     PP.Diag(Loc, diag::warn_max_tokens)
         << PP.getTokenCount() << (unsigned)MaxTokens;
diff --git a/clang/lib/Parse/Parser.cpp b/clang/lib/Parse/Parser.cpp
index 764d4e8e9d52..492eb058edfe 100644
--- a/clang/lib/Parse/Parser.cpp
+++ b/clang/lib/Parse/Parser.cpp
@@ -651,6 +651,10 @@ bool Parser::ParseTopLevelDecl(DeclGroupPtrTy &Result, bool IsFirstDecl) {
       }
     }
 
+    if (PP.getPPCallbacks()) {
+      PP.getPPCallbacks()->PragmaMaxTokensTotal(PP.getMaxTokensOverrideLoc(), PP.getTokenCount(), PP.getMaxTokens());
+    }
+
     // Late template parsing can begin.
     Actions.SetLateTemplateParser(LateTemplateParserCallback, nullptr, this);
     if (!PP.isIncrementalProcessingEnabled())
